Adds bookmarks sync in browsers (through an extension)

DT3264 opened the pull request:
Adds bookmarks sync in browsers (through an extension)

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/equetzal/huronOS-build-tools/pull/93).
* #164
* #139
* __->__ #93
* #158

equetzal reviewed (changes requested):
Ok, so I took a look into this PR and it seems something pretty early stage, but I'm really impressed of the work you're doing with this. Ok so I've not taken a look into the source code of the bookmark-bridge repo so I don't know how it works internally, but I'll do it as soon as I can. 

Now for this PR, I did a totally clean build, new debian, new system layers, everything was clean, and the first thing I noticed after building the chrome module is that the extension is not being loaded properly, but somehow it still working:
![Screenshot from 2023-06-11 21-54-50 copia](https://github.com/equetzal/huronOS-build-tools/assets/62947298/6dcb80af-6fca-4d30-9265-43702fa8c213)


Now, after launching both Firefox and Chromium it was noticeable that they were not working properly, in this case the directives URL was wrong so the system was trying to download the directives each minute which seemed to also trigger this extension with directives fallback to default and leading to ton of bookmarks:
![Screenshot from 2023-06-11 21-48-55](https://github.com/equetzal/huronOS-build-tools/assets/62947298/ac5974f6-6e24-4b5c-909f-a955184635e7)

After fixing the URL sync seemed to work, as the extensiones were added to the end, but this gives me to: Doesn't the extension should delete all bookmarks and replace them with the new ones? Why to append them?
![Screenshot from 2023-06-11 21-53-01 copia](https://github.com/equetzal/huronOS-build-tools/assets/62947298/46b11d1a-119a-46ad-ac39-a0f5f483a327)

Ok, so this were overall look around, this feature seems really good but I think we're needing to consider way more edge cases to better solve this issue.

DT3264 commented:
I took a look to this PR with a fresh mind and here it is my devlog on this one

- It doesn't show the "error loading extension" anymore. 
Seems like this one happened because of some automatic stuff in the extensions loader, mainly, after executing chromium, it would open itself with more arguments, one being `--load-extension /usr/share/chromium/extensions/EXTENSION_ID.json` but the load-extension arg expects an unpacked extension, the fix?, open chromium with the arguments shown on `chrome:version` and adding all of those arguments (but the load-extension one) to the .desktop file, and tadaaaa.

- The extension no longer repeats a bookmark if it lacks a slash
I learned that, for urls, they can be either a directory or a file, directories end with a slash, files do not. Now, adding a bookmark to a browser, they seem to decide internaly if an url is a directory or a file, and, for example, "omegaup.com" would get a slash at the end, this was the reason the issue of repeated  bookmarks, to solve it I remove the trailing slash (if it has one) on both the directives url to add and the current bookmark being compared against, this succesfully solves that issue.

- The required files for the extension now are downloaded from the bookmarks-bridge repo
I thought "Why do we need to have these files here if we already have them on github", so I removed them from this repo and instead fetch them directly from github. For the actual extension files, I have them as release files. And that repo now is in a clenaer state, although I made some force pushes to remove trace of some files and have a fresh start on the commit history.

One more thing, well, two things

- The bookmarks bar now automatically appears on firefox
- And the anoying "Import bookmark" button (I think of it as such) has been removed from firefox.

So, with the new changes I feel like this PR is both more valuable and closer to be ready than before.

DT3264 commented:
If testing this PR, I suggest to include on top #139 for a smoother firefox testing without having to rebuild the 1st layer for the firefox config file.

DT3264 commented:
Also, quick question, which do you think would be a good sleep time for the script so it doesn't send the bookmarks every 5 seconds to the browser?, I'm between 60 and 30 seconds, I feel that less might be too overwhelming for the browser, even if the data sent is small, yet I feel that whichever time is said, an already documented delay is going to be seen between the directives are updated and the new bookmarks displayed, but IDK, maybe I'm overthinking on something small. :sweat_smile: 

In the meantime I left 60 seconds as a default [here](https://github.com/huronOS/bookmark-bridge/blob/development/app/bookmarks-fetcher.py#L35).

equetzal commented:
> Also, @equetzal , which do you think would be a good sleep time for the script so it doesn't send the bookmarks to the browser every 5 seconds?, I'm between 30 and 60 seconds, I feel that less than 30 might be too overwhelming for the browser, even if the data sent is small, yet I feel that whichever time is said, an already documented delay is going to be seen between the directives being updated and the new bookmarks received on the browser, but IDK, maybe I'm overthinking on something small. 😅
> 
> In the meantime I left 60 seconds as a default [here](https://github.com/huronOS/bookmark-bridge/blob/development/app/bookmarks-fetcher.py#L35).

Sounds good a delay of 60s, but ideally it would be nice to have an observer that executes a callback just when file is updated. Not sure, but It should be a library or something that already watch files for changes and exec the callback.

equetzal commented:
Won't be able to test until 2 weeks. Want to test it before accepting, unless @RafagaBlanca or nico are able to test this by their side. But you might need to pressure them haha.

DT3264 commented:
> Sounds good a delay of 60s, but ideally it would be nice to have an observer that executes a callback just when file is updated. Not sure, but It should be a library or something that already watch files for changes and exec the callback.

You're right, I got stuck with the sleep since that was the first thing that got to my mind lol, I'll try to add the callback later

DT3264 commented:
I already added the file tracking, it was fun to work with, now the changes are shown as soon as the directives are redownlaoded! :open_mouth: 

I also moved all the extension file copying / downloading to their own setup scripts over the bookmark-bridge repo, and included the bookmark-bridge as submodule, that would keep things cleaner on both repos (at least I think it does haha).

DT3264 commented:
On the testing side, there's no rush on that, if anyone else wants to try it, thats good, if none and you're able to try it latter when you return, that's good too.

equetzal commented:
> I already added the file tracking, it was fun to work with, now the changes are shown as soon as the directives are redownlaoded! 😮
> 
> I also moved all the extension file copying / downloading to their own setup scripts over the bookmark-bridge repo, and included the bookmark-bridge as submodule, that would keep things cleaner on both repos (at least I think it does haha).

Nice update! 
Just, do you think we can move the submodule to another path?
`huronOS-build-tools/utils`, not sure of which name to use, it could be utils, tools, modules, apps¿, or something else, but I'm planning to reach a gold like this some day near:

```txt
utils/
    installer/
    make-iso/
    bookmark-bridge/
    sync-manager/
    hmount/
    auls/
    module-manager/
    etc...
```

It's part of the goal to modularize the system software that we've wrote.
I'm welcome to your opinion about which naming to use.

DT3264 commented:
> I'm welcome to your opinion about which naming to use.
What about modules?

I think that could be the most apropiated name since the folder contents seem to be closer to modules than tools or utils.

DT3264 commented:
@equetzal Pinging jic

equetzal commented:
> > I'm welcome to your opinion about which naming to use.
> 
> What about `modules`?
> 
> I think that could be the most apropiated name since the folder contents seem to be closer to modules than tools or utils.

Hmm I'm not sure about modules because we already have that name used for 'huronOS software modules', so people can confuse them don't you think?

DT3264 commented:
> > > I'm welcome to your opinion about which naming to use.
> > 
> > What about `modules`?
> > 
> > I think that could be the most apropiated name since the folder contents seem to be closer to modules than tools or utils.
> 
> Hmm I'm not sure about modules because we already have that name used for 'huronOS software modules', so people can confuse them don't you think? 

You're right, I forgot that.

What about artifacts or components?

I want to avoid utils or tools since I feel that the contents of the folder are going to be more than that.

equetzal commented:
> > > > I'm welcome to your opinion about which naming to use.
> > > 
> > > 
> > > What about `modules`?
> > > I think that could be the most apropiated name since the folder contents seem to be closer to modules than tools or utils.
> > 
> > 
> > Hmm I'm not sure about modules because we already have that name used for 'huronOS software modules', so people can confuse them don't you think?
> 
> You're right, I forgot that.
> 
> What about artifacts or components?
> 
> I want to avoid utils or tools since I feel that the contents of the folder are going to be more than that.

I think artifacts or packages could be a good fit.

DT3264 commented:
Okay, after adding the packages folder at the root of the project, I added the bookmark-bridge submodule to that path and updated the script execution as required :D

equetzal commented:
Let's make a second PR to update the docs to include the git clone command with the `--recurse-submodules` flag, as we'll be needing it for being sure that the new submodules are available in script runtime.

equetzal commented on software-modules/internet/chromium/chromium.sh:
```diff
@@ -20,12 +20,18 @@ TARGET_DIR="/run/initramfs/memory/system/huronOS/software/internet/"
-apt install --yes --no-install-recommends $NAME
+# TODO: When migrating to debian12, replace pip with python3-watchdog and remove the following two lines
+apt install --yes --no-install-recommends $NAME python-is-python3 pip
+pip install watchdog
```
If  both chromium and firefox needs watchdog why not to include it on the shared-libs instead?

DT3264 replied:
You're right, I forgot to move that, already moving

equetzal reviewed (approved):
Overall good, I've tested this PR and now the extension seems way more stable, it have some caveats, so I've opened some Issues on the repo of the extension:
https://github.com/huronOS/bookmark-bridge/issues/2
https://github.com/huronOS/bookmark-bridge/issues/3
I also commented some issues with the build scripts for the browser packages, nothing big. Just fix them before merging.

In overall, this extension makes all things that the initial bookmarks approach did, plus now we're able to add new bookmarks on the fly which is amazing!! 
Congrats on this achievement!!

![](https://64.media.tumblr.com/tumblr_m20dh0C4Uz1rrdywqo1_250.gif)

DT3264 commented on software-modules/internet/firefox/firefox.sh:
```diff
@@ -20,12 +20,18 @@ TARGET_DIR="/run/initramfs/memory/system/huronOS/software/internet/"
 cp ./$NAME.desktop /usr/share/applications/
 
+# Prepare the bookmark-bridge extension
+./../../packages/bookmark-bridge/firefox/setup-extension.sh
```
Done

DT3264 commented on software-modules/internet/chromium/chromium.sh:
```diff
@@ -20,12 +20,18 @@ TARGET_DIR="/run/initramfs/memory/system/huronOS/software/internet/"
 cp ./$NAME.desktop /usr/share/applications/
 
+# Prepare the bookmark-bridge extension
+./../../packages/bookmark-bridge/chromium/setup-extension.sh
```
Done

equetzal commented:
Follow up to this comment, please: 
https://github.com/equetzal/huronOS-build-tools/pull/139#issuecomment-1658917803

equetzal commented:
@DT3264 I've tested this again. LGTM!
Feel free to merge this PR :D 
Congrats!!
